File: w01/JunHyeok/2798_2.py
n, target = map(int, input().split())
arr = list(map(int, input().split()))
res = 0
lst = []

def func(start, idx):
    global res
    global vect
    global target
    global arr
    global n

    if idx == 3:
        sum = 0
        for i in lst:
            sum += i
            # print(str(i) + " ", end = " ") 각 단계별로 어떤 조합을 사용하는지 보고싶으면 주석을 해제하세요
        # print("Sum : " + str(sum)) 각 단계별로 어떤 조합을 사용하는지 보고싶으면 주석을 해제하세요
        if sum <= target:
            res = max(res, sum)
        return

    for i in range(start, n):
        lst.append(arr[i])
        func(i + 1, idx + 1)
        lst.pop()

func(0, 0)
print(res)

File: w01/JunHyeok/2798_2.cpp
#include <iostream>
#include <vector>

using namespace std;

int n, target;
int arr[100] = {0,};
int res = 0;
int sum = 0;
vector<int> vect;

void func(int start, int idx) {
    if (idx == 3) {
        for (auto i : vect) {
//            cout << i << " "; 각 단계별로 어떤 조합을 사용하는지 보고싶으면 주석을 해제하세요
            sum += i;
        }
//        cout << "Sum : " << sum << "\n"; 각 단계별로 어떤 조합을 사용하는지 보고싶으면 주석을 해제하세요
        if (sum <= target) {
            res = max(res,sum);
        }
        sum = 0;
        return;
    }

    for (int i=start; i<n; i++) {
        vect.push_back(arr[i]);
        func(i + 1, idx + 1);
        vect.pop_back();
    }
}

int main() {

    cin >> n >> target;

    for (int i=0; i<n; i++) {
        cin >> arr[i];
    }

    func(0,0);

    cout << res;
    return 0;
}
